mpc-core/src/protocols/rep3/yao.rs

```diff
@@ -0,0 +1,480 @@
+//! Yao
+//!
+//! This module contains operations with Yao's garbled circuits
+
+pub mod circuits;
+pub mod evaluator;
+pub mod garbler;
+
+use super::{
+    network::{IoContext, Rep3Network},
+    IoResult, Rep3BigUintShare, Rep3PrimeFieldShare,
+};
+use crate::protocols::rep3::id::PartyID;
+use ark_ff::{PrimeField, Zero};
+use fancy_garbling::{BinaryBundle, WireLabel, WireMod2};
+use itertools::Itertools;
+use num_bigint::BigUint;
+use rand::{CryptoRng, Rng};
+use scuttlebutt::Block;
+
+/// A structure that contains both the garbler and the evaluators wires
+pub struct GCInputs<F> {
+    /// The wires of the garbler. These represent random keys x_0
+    pub garbler_wires: BinaryBundle<F>,
+    /// The wires of the evaluator. These represent the keys x_c = x_0 xor delta * val
+    pub evaluator_wires: BinaryBundle<F>,
+    /// The delta used for encoding known to the garbler
+    pub delta: F,
+}
+
+/// This struct contains some useful utility functions for garbled circuits.
+pub struct GCUtils {}
+
+impl GCUtils {
+    pub(crate) fn garbled_circuits_error<G, T>(input: Result<T, G>) -> IoResult<T> {
+        input.or(Err(std::io::Error::new(
+            std::io::ErrorKind::Other,
+            "Garbled Circuit failed",
+        )))
+    }
+
+    fn receive_block_from<N: Rep3Network>(network: &mut N, id: PartyID) -> IoResult<Block> {
+        let data: Vec<u8> = network.recv(id)?;
```

Comment:
Think sending single blocks over our current network is also not great tbh, so I would prefer having the above variant.